lib/elixir/lib/kernel.ex

```diff
@@ -3069,6 +3076,39 @@ defmodule Kernel do
     end
   end
 
+  # Called as ensure_evaled(list, {0, []}). Note acc is reversed.
+  def ensure_evaled(list, acc) do
```

Comment:
I just realized this is changing the semantics of the code too. Take this:

```elixir
iex(1)> a = 1
1
iex(2)> [a = 2, a]
[2, 1]
```

Now because the execution is moved up, the value of the second a will be `2` instead of 1. You probably want to make the acc a list of `vars to values` and compile that to a single match instead.